packages/core/src/tools/utils.ts

```diff
@@ -156,24 +156,32 @@ export function noop() {}
  *
  * Note: this still assumes that JSON.stringify is correct.
  */
-export function jsonStringify(
-  value: unknown,
-  replacer?: Array<string | number>,
-  space?: string | number
-): string | undefined {
+export function jsonStringify(value: unknown, space?: string | number): string | undefined {
   if (typeof value !== 'object' || value === null) {
     return JSON.stringify(value)
   }
 
-  // Note: The order matter here. We need to detach toJSON methods on parent classes before their
+  // Note: The order matters here. We need to detach toJSON methods on parent classes before their
   // subclasses.
   const restoreObjectPrototypeToJson = detachToJsonMethod(Object.prototype)
   const restoreArrayPrototypeToJson = detachToJsonMethod(Array.prototype)
   const restoreValuePrototypeToJson = detachToJsonMethod(Object.getPrototypeOf(value))
   const restoreValueToJson = detachToJsonMethod(value)
 
+  // Line below can be removed when migrating to TS4.8+
+  const isValidObject = (v: any): v is object => typeof v === 'object' && v !== null
+  const getCyclicReplacer = <T>() => {
```

Comment:
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**🥜 nitpick:** ‏the type parameter `T` isn't useful here, and could be replaced by `unknown` in the argument list

<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**🥜 nitpick:** ‏FMU you create a closure just to keep the `circularReferenceChecker` variable private. IMO this is not worth it and the following would be a bit easier to read:

```ts
const circularReferenceChecker = createCircularReferenceChecker()
const cyclicReplacer = (_key: string, value: unknown) => {
  if (isValidObject(value) && circularReferenceChecker.hasAlreadyBeenSeen(value)) {
    return '<warning: cyclic reference not serialized>'
  }
  return value
}
```